File: game/src/entities/Player.cpp
#include "Player.h"
#include "graphics/AssetManager.h"
#include "graphics/Renderer.h"
#include "math/Vec2.h"
#include "physics/CircleCollider.h"
#include "physics/PhysicsManager.h"
#include "physics/Rigidbody.h"
#include <raylib.h>

Player::Player(Vec2 position)
{
	this->tag = "player";

	this->position = position;
	this->scale = 4.0f;
	this->rotation = 0.0f;

	this->speed = 1500.0f;

	this->collider = CircleCollider(position, 10.0f);
	this->rigidbody = Rigidbody(1.0f, 7.5f, false);

	this->origin = Vec2(0.5f, 0.5f);
}

Player::~Player()
{
	PhysicsManager::unregisterBody(id);
	UnloadTexture(texture);
}

void Player::load()
{
	texture = AssetManager::loadTexture("assets/sprites/player.png");
	PhysicsManager::registerBody(id, &rigidbody, &position);
}

void Player::update()
{
	// get input
	Vec2 input = Vec2::zero();
	if (IsKeyDown(KEY_D)) input.x += 1.0f;
	if (IsKeyDown(KEY_A)) input.x -= 1.0f;
	if (IsKeyDown(KEY_S)) input.y += 1.0f;
	if (IsKeyDown(KEY_W)) input.y -= 1.0f;

	// apply forces
	rigidbody.applyForce(input * speed);
	collider.setPosition(position);
}

void Player::draw()
{
	Renderer::drawTexture(texture, position, scale, rotation, origin, WHITE);
}

File: examples/game/entities/player.go
package entities

import (
	"froglib/assets"
	. "froglib/components"
	"froglib/input"
	. "froglib/math"

	"github.com/hajimehoshi/ebiten/v2"
)

type Player struct {
	*Transform
	*Rigidbody
	sprite *Sprite

	speed  float64
	origin Vec2
}

func NewPlayer(x, y float64) *Player {
	player := Player{
		Transform: NewTransform(NewVec2(0.0, 0.0), 2.0, 0.0),
		Rigidbody: NewRigidbody(1.0, 0.5, false),
		sprite:    NewSprite(assets.GetImage("player")),

		speed:  1.0,
		origin: NewVec2(0.5, 0.5),
	}

	return &player
}

func (p *Player) Update() {
	// handle movement
	dir := input.GetVector(ebiten.KeyA, ebiten.KeyD, ebiten.KeyW, ebiten.KeyS)
	p.ApplyForce(dir.Scale(p.speed))

	p.Integrate(&p.Position)
}

func (p *Player) Draw(screen *ebiten.Image) {
	options := &ebiten.DrawImageOptions{}
	p.Apply(options, p.sprite.Size(), p.origin)

	p.sprite.Draw(screen, options)
}
